bullet/src/JointFeatures.cc

```diff
@@ -0,0 +1,478 @@
+	    hinge->getFrameOffsetA().getBasis().getColumn(2);
+
+	  math::Vector3 globalAxis(vec[0], vec[1], vec[2]);
+
+	  if (this->links.find(jointInfo->childLinkId) != this->links.end())
+          {
+	    btRigidBody *childLink = this->links.at(jointInfo->childLinkId)->link.get();
+	    btVector3 aux = childLink->getTotalTorque();
+	    math::Vector3 angularTorque(aux[0], aux[1], aux[2]);
+	    result += globalAxis.Dot(angularTorque);
+	  }
+	  if (this->links.find(jointInfo->parentLinkId) != this->links.end())
+	  {
+	    btRigidBody *parentLink =
+	      this->links.at(jointInfo->parentLinkId)->link.get();
+	    btVector3 aux = parentLink->getTotalTorque();
+	    math::Vector3 angularTorque(aux[0], aux[1], aux[2]);
+	    result -= globalAxis.Dot(angularTorque);
+	  }
+	}
+	else
+	{
+	  ignerr << "Corrupted joint at index:" << _id.id << "\n";
+	}
+      }
+      break;
+    default:
+      ignwarn << "Not a valid getJointForce type: " << jointType << "\n";
+      break;
+    }
+  }
+  return result;
+}
+
+/////////////////////////////////////////////////
+Pose3d JointFeatures::GetJointTransform(const Identity &_id) const
+{
+  (void) _id;
+  ignwarn << "Dummy function GetJointTransform\n";
+  return Pose3d();
+}
+
+/////////////////////////////////////////////////
+void JointFeatures::SetJointPosition(
+    const Identity &_id, const std::size_t _dof, const double _value)
+{
+  (void) _id;
+  (void) _dof;
+  (void) _value;
+  ignwarn << "Dummy function SetJointPosition\n";
+}
+
+/////////////////////////////////////////////////
+void JointFeatures::SetJointVelocity(
+    const Identity &_id, const std::size_t _dof, const double _value)
+{
+  (void) _id;
+  (void) _dof;
+  (void) _value;
+  ignwarn << "Dummy SetJointVelocity\n";
+}
+
+/////////////////////////////////////////////////
+void JointFeatures::SetJointAcceleration(
+    const Identity &_id, const std::size_t _dof, const double _value)
+{
+  (void) _id;
+  (void) _dof;
+  (void) _value;
+  ignwarn << "Dummy SetJointAcceleration\n";
+}
+
+/////////////////////////////////////////////////
+void JointFeatures::SetJointForce(
+    const Identity &_id, const std::size_t _dof, const double _value)
+{
+  (void) _dof;
+
+  if (this->joints.find(_id.id) != this->joints.end())
+  {
+    const JointInfoPtr &jointInfo = this->joints.at(_id.id);
+    const int jointType = jointInfo->constraintType;
+    switch(jointInfo->constraintType)
+    {
+      case static_cast<int>(::sdf::JointType::REVOLUTE) :
+      {
+	btHingeAccumulatedAngleConstraint* hinge =
+	  static_cast<btHingeAccumulatedAngleConstraint*>(jointInfo->joint.get());
+	if (hinge)
+	{
+	  // Limit the max torque applied to avoid abrupt changes in the
+	  // angular position of the joint and losing the angle reference
+	  // TO-DO (blast545): this limitation should be based on angular speed
+	  // as this breaks the PID controller when setting high values
+	  const double thresholdValue = std::max(std::min(_value, 0.1), -0.1);
```

Comment:
@azeey Mmm, we added this because increasing the speed of a revolute joint up to a certain point made the internal controller to lose track of the angle. We noted it as a known issue, but IIRC it only affected aerial vehicles. I'm OK with removing the threshold as long as the "known issue" remains visible. WDYT?

@Lobotuerk Also WDYT

Reply:
I agree, lets remove the threshold, and point out in the issue to the bullet's codebase part that has that "if speed > cost, free angle"

Reply:
Which controller are you referring to?

Reply:
It's not a controller, it's the internal "angle tracker" that bullet uses. You can see the problem we are refering to here: https://pybullet.org/Bullet/BulletFull/btHingeConstraint_8cpp_source.html#l00318. If there's a change in the angle bigger than the internal threshold within two simulation times, the angle tracker resets the count to the absolute current angle.